src/lib/support/BufferWriter.h

```diff
@@ -0,0 +1,185 @@
+    BufferWriterBase & Put(const void * buf, size_t len)
+    {
+        size_t available = Available();
+
+        memmove(mBuf + mNeeded, buf, available < len ? available : len);
+
+        mNeeded += len;
+
+        return *this;
+    }
+
+    BufferWriterBase & Skip(size_t len)
+    {
+        mNeeded += len;
+        return *this;
+    }
+
+    /// Append a single byte
+    BufferWriterBase & Put(uint8_t c)
+    {
+        if (mNeeded < mSize)
+        {
+            mBuf[mNeeded] = c;
+        }
+        ++mNeeded;
+        return *this;
+    }
+
+    // write an integer into a buffer, in an endian specific way
+
+    BufferWriterBase & Put8(uint8_t c) { return static_cast<Derived *>(this)->Put(c); }
+    BufferWriterBase & Put16(uint16_t x) { return static_cast<Derived *>(this)->EndianPut(x, sizeof(x)); }
+    BufferWriterBase & Put32(uint32_t x) { return static_cast<Derived *>(this)->EndianPut(x, sizeof(x)); }
+    BufferWriterBase & Put64(uint64_t x) { return static_cast<Derived *>(this)->EndianPut(x, sizeof(x)); }
+
+    /// Number of bytes required to satisfy all calls to Put() so far
+    size_t Needed() const { return mNeeded; }
+
+    /// Number of bytes still available for writing
+    size_t Available() const { return mSize < mNeeded ? 0 : mSize - mNeeded; }
+
+    /// Whether the input fit in the buffer
+    bool Fit() const
+    {
+        size_t _;
+        return Fit(_);
+    }
+
+    /// Returns whether the input fit in the buffer, outputs what was actually written
+    bool Fit(size_t & actually_written) const
+    {
+        actually_written = mSize >= mNeeded ? mNeeded : mSize;
+        return mSize >= mNeeded;
+    }
+
+    /// Size of the output buffer
+    size_t Size() const { return mSize; }
+
+    uint8_t * Buffer() { return mBuf; }
+    const uint8_t * Buffer() const { return mBuf; }
+
+private:
+    friend Derived;
+
+    uint8_t * mBuf;
+    size_t mSize;
+    size_t mNeeded;
+
+    BufferWriterBase(uint8_t * buf, size_t len) : mBuf(buf), mSize(len), mNeeded(0) {}
+    BufferWriterBase(const BufferWriterBase & other) = default;
+    BufferWriterBase & operator=(const BufferWriterBase & other) = default;
+};
+
+namespace LittleEndian {
+
+class BufferWriter : public BufferWriterBase<BufferWriter>
+{
+public:
+    BufferWriter(uint8_t * buf, size_t len) : BufferWriterBase<BufferWriter>(buf, len) {}
+    BufferWriter(const BufferWriter & other) = default;
+    BufferWriter & operator=(const BufferWriter & other) = default;
+
+    BufferWriter & EndianPut(uint64_t x, size_t size)
+    {
+        while (size-- > 0)
+        {
+            uint8_t c = x & 0xff;
+            Put(c);
+            x >>= 8;
+        }
+        return *this;
+    }
+
+    // allows treating of a buffer stream to a different endianess
+    template <typename T>
+    static BufferWriter StartingFrom(BufferWriterBase<T> & other)
+    {
+        if (!other.Fit())
+        {
+            return BufferWriter(nullptr, 0);
```

Comment:
This loses information.. The new write "fits" until something is written to it, right?

Reply:
Updated and added unit tests.